src/hotspot/share/gc/shenandoah/shenandoahRuntime.hpp

```diff
@@ -37,6 +37,9 @@ class ShenandoahRuntime : public AllStatic {
   static void arraycopy_barrier_narrow_oop(narrowOop* src, narrowOop* dst, size_t length);
 
   static void write_ref_field_pre(oopDesc* orig, JavaThread* thread);
+  static void pre_barrier(JavaThread* thread, oopDesc* orig) {
+    write_ref_field_pre(orig, thread);
+  }
```

Comment:
So, why not export `write_ref_field_pre`, instead of introducing this new method? Style/cleanliness, or something else? I am asking, because every time we add a new stub here, we would need to record it in `AOTCache` tables for Leyden benefit.

Reply:
It's about the argument ordering. Graal expects the Thread* to be prependend, while other JITs call it with the Thread* appended. I guess we could change other JIT calls to also prepend the thread, or change the interface to not pass the Thread* at all. I chose to follow G1 and export both variants.

Reply:
Oh, so this matches `JVMCIRuntime::write_barrier_pre` for G1 (weird place to have it, but oh well). 

Does Graal needs the `Thread*` argument?

I think this method is only called when SATB buffer is full. So the performance of this method is likely not affected by getting the current thread down in caller. So I think it would be more straight-forward to sharpen `ShenandoahRuntime::write_ref_field_pre` by dropping `Thread*` and then exporting that. Maybe also under the `SR::write_barrier_pre` name to be even more consistent for everything else.

Maybe @JohnTortugo wants to clean up more mess in C2 related to this :)

Reply:
Graal does not need the Thread* argument, but the runtime code behind write_ref_pre() currently uses it. I agree, it does not look performance critical to pass it through. However, getting rid of it seems to blow the scope of this PR. I'd rather do this as a follow-up.

Reply:
Actually, I'd probably add the new entry for Graal without the Thread* argument now, and fix the others in a follow-up. Otherwise we need to deal with it on the Graal side again later once we change the entry points.

Reply:
OK, but that follow-up risks changing the JVMCI interface _again_? How about we introduce:

```
static void write_barrier_pre(oopDesc* pre_val) {
  write_ref_field_pre(pre_val, JavaThread::current());
}
```

...and then the follow-up purges the old `write_ref_field_pre`? The implementation might need to be in `.cpp`.